src/main/java/com/hivemq/cli/mqtt/test/Mqtt3FeatureTester.java

```diff
@@ -0,0 +1,536 @@
+        final Mqtt3Client subscriber = buildClient();
+        final String topic = TopicUtils.generateTopicUUID(maxTopicLength);
+        final byte[] payload = qos.toString().getBytes();
+
+        subscriber.toBlocking().connect();
+        publisher.toBlocking().connect();
+
+        final CountDownLatch countDownLatch = new CountDownLatch(tries);
+        final AtomicInteger totalReceived = new AtomicInteger(0);
+
+        try {
+            subscriber.toAsync().subscribeWith()
+                    .topicFilter(topic)
+                    .qos(qos)
+                    .callback(publish -> {
+                        if (publish.getQos() == qos
+                                && Arrays.equals(publish.getPayloadAsBytes(), payload)) {
+                            totalReceived.incrementAndGet();
+                            countDownLatch.countDown();
+                        }
+                    })
+                    .send()
+                    .join();
+        } catch (final Exception ex) {
+            Logger.error(ex, "Could not subscribe with QoS {}", qos.ordinal());
+        }
+
+        final long before = System.nanoTime();
+
+        for (int i = 0; i < tries; i++) {
+            try {
+                publisher.toAsync().publishWith()
+                        .topic(topic)
+                        .qos(qos)
+                        .payload(payload)
+                        .send();
+            } catch (final Exception ex) {
+                Logger.error("Could not publish with QoS {}", qos.ordinal());
+            }
+        }
+
+        try {
+            countDownLatch.await(timeOut, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            Logger.error("Interrupted while waiting for QoS {} publish to arrive at subscriber", qos.ordinal());
+        }
+
+        final long after = System.nanoTime();
+        final long timeToComplete = after - before;
+
+        disconnectIfConnected(publisher, subscriber);
+
+        return new QosTestResult(totalReceived.get(), timeToComplete);
+    }
+
+    public @NotNull PayloadTestResults testPayloadSize(final int maxSize) {
+        final Mqtt3Client subscriber = buildClient();
+        final Mqtt3Client publisher = buildClient();
+        final List<Tuple<Integer, TestResult>> testResults = new LinkedList<>();
+        final String topic = (maxTopicLength == -1 ? TopicUtils.generateTopicUUID() : TopicUtils.generateTopicUUID(maxTopicLength));
+        final Mqtt3Publishes publishes = subscriber.toBlocking().publishes(MqttGlobalPublishFilter.SUBSCRIBED);
+        final String oneByte = "a";
+        int top = maxSize;
+        int bottom = 0;
+        int mid = -1;
+
+        subscriber.toBlocking().connect();
+        publisher.toBlocking().connect();
+
+        subscriber.toBlocking().subscribeWith()
+                .topicFilter(topic)
+                .qos(MqttQos.AT_LEAST_ONCE)
+                .send();
+
+        // Binary search the payload size
```

Comment:
I think you should test the maximum payload size first and only if it failed use this binary search of the limit

Reply:
Sounds good. 👍

Reply:
I think, I`ll test the maximums on all tests first and only if that fails I will do the binary searches.